### PR TITLE
Updates to SyncPoller

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

Fixes https://github.com/Azure/azure-sdk-for-java/issues/37529

Makes the `SyncPoller` implementations `SimpleSyncPoller` and `SyncOverAsyncPoller` consistent and updates the documentation for `SyncPoller` to clarify behaviors.

`SyncOverAsyncPoller`'s implementation changed to match `SimpleSyncPoller`. `waitForCompletion` and `getFinalResult` methods with `Duration timeout` now will throw a `RuntimeException` wrapping a `TimeoutException` if any poll operation takes longer than the timeout or will return the last received polling response received during the polling timeout.

Documentation now clarifies that the `Duration timeout` parameter has two usages. First is triggering a timeout exception if any poll operation takes longer than the timeout. Second is the `Duration` determines how long polling will run for until a value is returned and clarifies for the methods taking a `LongRunningOperationStatus` that the returned response's status may not match what was requested or won't be a completion status.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollingUtil.java [hunk 1]
```diff
@@ -46,18 +45,27 @@ static <T> PollResponse<T> pollingLoop(PollingContext<T> pollingContext, Duratio
...
+                if (intermediatePollResponse.getStatus().equals(statusToWaitFor)) {
+                    return intermediatePollResponse;
+                } else {
+                    throw LOGGER.logExceptionAsError(new RuntimeException(
+                        new TimeoutException("Polling didn't complete before the timeout period.")));
+                }
             }
-            if (statusToWaitFor != null && intermediatePollResponse.getStatus().equals(statusToWaitFor)) {
+
+            if (intermediatePollResponse.getStatus().equals(statusToWaitFor)) {
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollingUtil.java:61]: We've got the same problem here as we might poll until completion if we didn't poll the service when the operation reached the expected state and had moved on from that state. However, we don't have a solution to this and we'll continue polling to reach the completion state.

File: sdk/core/azure-core/src/main/java/com/azure/core/util/polling/SyncPoller.java [hunk 2]
```diff
@@ -38,48 +39,107 @@ public interface SyncPoller<T, U> {
...
+     *
+     * @param timeout the duration to wait for polling completion.
+     * @return the final result of the long-running operation if there is one.
+     * @throws NullPointerException If {@code timeout} is null.
+     * @throws IllegalArgumentException If {@code timeout} is zero or negative.
+     * @throws RuntimeException If polling doesn't complete before the {@code timeout} elapses.
+     * ({@link RuntimeException#getCause()} should be a {@link TimeoutException}).
+     */
+    default U getFinalResult(Duration timeout) {
+        return getFinalResult();
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/polling/SyncPoller.java:134]: Should we instead do this to honor the timeout value?

```java
waitForCompletion(timeout);
return getFinalResult();
```

File: sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollingUtil.java [hunk 3]
```diff
@@ -46,18 +45,27 @@ static <T> PollResponse<T> pollingLoop(PollingContext<T> pollingContext, Duratio
...
         };
+
         while (!intermediatePollResponse.getStatus().isComplete()) {
             long elapsedTime = System.currentTimeMillis() - startTime;
-            if (timeBound ?  elapsedTime >= timeoutInMillis : false) {
+            if (timeBound && elapsedTime >= timeoutInMillis) {
                 scheduler.shutdown();
-                return intermediatePollResponse;
+                if (intermediatePollResponse.getStatus().equals(statusToWaitFor)) {
+                    return intermediatePollResponse;
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollingUtil.java:54]: Yeah, this is definitely a more interesting case. I agree that the least confusing, and tricky, option would be just returning at the end of the timeout or if the request state is returned or if a terminal state is returned.

**Review by srnagar — APPROVED** [review]

File: sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollingUtil.java [hunk 4]
```diff
@@ -30,56 +32,70 @@
...
-                return intermediatePollResponse;
-            }
-            final ScheduledFuture<?> pollOp =
-                scheduler.schedule(pollOpRunnable, getDelay(intermediatePollResponse, pollInterval).toMillis(), TimeUnit.MILLISECONDS);
-            try {
-                if (timeBound) {
-                    pollOp.get(timeoutInMillis - elapsedTime, TimeUnit.MILLISECONDS);
+
+        boolean firstPoll = true;
+        final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollingUtil.java:49]: Probably, but I'd leave that to another PR which can specifically focus on that.

**Pull request merged**